ntp-daemon/src/peer.rs

```diff
@@ -1,177 +1,272 @@
+    last_send_timestamp: Option<NtpTimestamp>,
+
+    /// Instant last poll message was sent (used for timing the wait)
+    last_poll_sent: Instant,
+
+    /// Number of resets that this peer has performed
+    reset_epoch: ResetEpoch,
+}
+
+impl<C> PeerProcess<C>
+where
+    C: 'static + NtpClock + Send,
+{
+    /// Set the next deadline for the poll interval based on current state
+    fn update_poll_wait(&self, poll_wait: &mut Pin<&mut Sleep>, system_snapshot: SystemSnapshot) {
+        let poll_interval = self
+            .peer
+            .current_poll_interval(system_snapshot)
+            .as_system_duration();
+
+        poll_wait
+            .as_mut()
+            .reset(self.last_poll_sent + poll_interval);
+    }
+
+    async fn handle_poll(&mut self, poll_wait: &mut Pin<&mut Sleep>) {
+        let system_snapshot = *self.system_snapshots.read().await;
+        let packet = self.peer.generate_poll_message(system_snapshot);
+
+        // Sent a poll, so update waiting to match deadline of next
+        self.last_poll_sent = Instant::now();
+        self.update_poll_wait(poll_wait, system_snapshot);
+
+        match self.clock.now() {
+            Err(e) => {
+                // we cannot determine the origin_timestamp
+                panic!("`clock.now()` reported an error: {:?}", e)
+            }
+            Ok(ts) => {
+                self.last_send_timestamp = Some(ts);
+            }
+        }
+
+        if let Err(e) = self.socket.send(&packet.serialize()).await {
+            warn!(error = debug(e), "poll message could not be sent");
+        }
+    }
+
+    async fn handle_packet(
+        &mut self,
+        poll_wait: &mut Pin<&mut Sleep>,
+        packet: NtpHeader,
+        send_timestamp: NtpTimestamp,
+        recv_timestamp: NtpTimestamp,
+    ) {
+        let ntp_instant = NtpInstant::now();
+
+        let system_snapshot = *self.system_snapshots.read().await;
+        let result = self.peer.handle_incoming(
+            system_snapshot,
+            packet,
+            ntp_instant,
+            self.config.frequency_tolerance,
+            send_timestamp,
+            recv_timestamp,
+        );
+
+        // Handle incoming may have changed poll interval based on message, respect that change
+        self.update_poll_wait(poll_wait, system_snapshot);
+
+        let system_poll = system_snapshot.poll_interval.as_duration();
+        let accept = self.peer.accept_synchronization(
+            ntp_instant,
+            self.config.frequency_tolerance,
+            self.config.distance_threshold,
+            system_poll,
+        );
+
+        if let Some(msg) = prepare_msg_for_system(self.index, self.reset_epoch, accept, result) {
+            self.msg_for_system_sender.send(msg).await.ok();
+        }
+    }
 }
 
 #[instrument(skip(clock, config, system_snapshots, reset))]
 pub async fn start_peer<A: ToSocketAddrs + std::fmt::Debug, C: 'static + NtpClock + Send>(
```

Comment:
I'd suggest turning this into a `spawn()` static method on the `PeerProcess` type.